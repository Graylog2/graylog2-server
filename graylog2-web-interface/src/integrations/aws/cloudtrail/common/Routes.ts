/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { qualifyUrls } from 'routing/Routes';

const AWSCloudTrailRoutes = {
  INTEGRATIONS: {
    AWSCloudTrail: {
      ACTIVITYAPI: {
        index: '/integrations/cloudtrail',
      },
    },
  },
};

const ApiRoutes = {
  INTEGRATIONS: {
    AWSCloudTrail: {
      SAVE_INPUT: '/plugins/org.graylog.integrations/cloudtrail/inputs?setup_wizard=true',
      CHECK_CREDENTIALS: '/plugins/org.graylog.integrations/cloudtrail/check_credentials',
      GET_AWS_REGIONS: '/plugins/org.graylog.integrations/cloudtrail/getawsregions',
    },
  },
};

const DocsRoutes = {
  INTEGRATIONS: {
    AWSCloudTrail: {
      GRAYLOG_AWSCloudTrail_ACTIVITY_LOG_INPUT:
        'integrations/inputs/graylog_AWSCloudTrail_ActivityLog_Input.html#graylog-AWSCloudTrail-activitylog-input',
    },
  },
};

export default {
  ...qualifyUrls(AWSCloudTrailRoutes),
  unqualified: AWSCloudTrailRoutes,
};

export { DocsRoutes, ApiRoutes };
