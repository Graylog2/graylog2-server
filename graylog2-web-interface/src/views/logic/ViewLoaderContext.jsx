// @flow strict
import * as React from 'react';

export type ViewLoaderContextType = string => mixed;

const ViewLoaderContext = React.createContext<?ViewLoaderContextType>();

export default ViewLoaderContext;
