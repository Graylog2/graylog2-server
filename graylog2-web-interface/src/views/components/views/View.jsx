// @flow strict
import * as React from 'react';
import PropTypes from 'prop-types';
// $FlowFixMe: imports from core need to be fixed in flow
import { Link } from 'react-router';

import Routes from 'routing/Routes';
import connect from 'stores/connect';
import { EntityListItem } from 'components/common';
import StoreProvider from 'injection/StoreProvider';
import UserTimezoneTimestamp from 'views/components/common/UserTimezoneTimestamp';
import withPluginEntities from 'views/logic/withPluginEntities';

const CurrentUserStore = StoreProvider.getStore('CurrentUser');

const formatTitle = (title, id, disabled = false) => (disabled
  ? <h2>{title}</h2>
  : <Link to={Routes.pluginRoute('DASHBOARDS_VIEWID')(id)}>{title}</Link>);

const _OwnerTag = ({ owner, currentUser }) => {
  if (!owner || owner === currentUser.username) {
    return <span>Last saved</span>;
  }

  return <span>Shared by {owner}, last saved</span>;
};

_OwnerTag.propTypes = {
  owner: PropTypes.string.isRequired,
  currentUser: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
};

const OwnerTag = connect(_OwnerTag, { currentUser: CurrentUserStore }, ({ currentUser }) => ({ currentUser: currentUser.currentUser }));

// eslint-disable-next-line react/prop-types
const Description = ({ description, owner, createdAt }) => (
  <React.Fragment>
    <div>{description || <i>No description given.</i>}</div>
    <div style={{ color: 'darkgray' }}><OwnerTag owner={owner} /> at <UserTimezoneTimestamp dateTime={createdAt} /></div>
  </React.Fragment>
);

type Plugin = {
  name: string,
  url: string,
};

type RequirementsList = { [string]: Plugin };

const missingRequirements = (requires, requirementsProvided): RequirementsList => Object.entries(requires)
  .filter(([require]) => !requirementsProvided.includes(require))
  .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {});

const isMissingRequirements = (requires, requirementsProvided) => Object.keys(missingRequirements(requires, requirementsProvided)).length > 0;

type RequirementsProps = {
  requires: RequirementsList,
  requirementsProvided: Array<string>,
};

type MissingRequirementProps = {
  plugin: Plugin,
};

const MissingRequirement = ({ plugin }: MissingRequirementProps) => (
  <a href={plugin.url} target="_blank" rel="noopener noreferrer"><strong>{plugin.name}</strong></a>
);

const Requirements = ({ requires, requirementsProvided }: RequirementsProps) => {
  const missing = missingRequirements(requires, requirementsProvided);
  return Object.keys(missing).length > 0
    ? (
      <h5>
        Missing requirement(s): {Object.values(missing)
        // $FlowFixMe: plugin is of type Plugin, not mixed.
        .map((plugin: Plugin) => <MissingRequirement key={plugin.name} plugin={plugin} />)}
      </h5>
    )
    : null;
};

const View = ({ children, id, title, summary, description, owner, requires, createdAt, requirementsProvided }) => (
  <EntityListItem title={formatTitle(title, id, isMissingRequirements(requires, requirementsProvided))}
                  titleSuffix={summary}
                  description={(
                    <React.Fragment>
                      <Description description={description} owner={owner} createdAt={createdAt} />
                      <Requirements requires={requires} requirementsProvided={requirementsProvided} />
                    </React.Fragment>
                  )}
                  actions={children} />
);

View.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  summary: PropTypes.string,
  description: PropTypes.string,
  owner: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  requires: PropTypes.objectOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
  })).isRequired,
  requirementsProvided: PropTypes.arrayOf(PropTypes.string).isRequired,
};

View.defaultProps = {
  children: null,
  title: 'Unnamed View',
  summary: null,
  description: null,
};

export default withPluginEntities(View, { requirementsProvided: 'views.requires.provided' });
