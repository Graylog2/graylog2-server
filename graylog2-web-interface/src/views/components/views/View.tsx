/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import { Link } from 'components/graylog/router';
import Routes from 'routing/Routes';
import { EntityListItem } from 'components/common';
import CurrentUserContext from 'contexts/CurrentUserContext';
import UserTimezoneTimestamp from 'views/components/common/UserTimezoneTimestamp';
import withPluginEntities from 'views/logic/withPluginEntities';

const formatTitle = (title, id, disabled = false) => (disabled
  ? <h2>{title}</h2>
  : <Link to={Routes.pluginRoute('DASHBOARDS_VIEWID')(id)}>{title}</Link>);

const OwnerTag = ({ owner }) => {
  const currentUser = useContext(CurrentUserContext);

  if (!owner || owner === currentUser?.username) {
    return <span>Last saved</span>;
  }

  return <span>Shared by {owner}, last saved</span>;
};

OwnerTag.propTypes = {
  owner: PropTypes.string.isRequired,
};

// eslint-disable-next-line react/prop-types
const Description = ({ description, owner, createdAt }) => (
  <>
    <div>{description || <i>No description given.</i>}</div>
    <div style={{ color: 'darkgray' }}><OwnerTag owner={owner} /> at <UserTimezoneTimestamp dateTime={createdAt} /></div>
  </>
);

type Plugin = {
  name: string,
  url: string,
};

type RequirementsList = { [key: string]: Plugin };

const missingRequirements = (requires, requirementsProvided): RequirementsList => Object.entries(requires)
  .filter(([require]) => !requirementsProvided.includes(require))
  .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {});

const isMissingRequirements = (requires, requirementsProvided) => Object.keys(missingRequirements(requires, requirementsProvided)).length > 0;

type RequirementsProps = {
  requires: RequirementsList,
  requirementsProvided: Array<string>,
};

type MissingRequirementProps = {
  plugin: Plugin,
};

const MissingRequirement = ({ plugin }: MissingRequirementProps) => (
  <a href={plugin.url} target="_blank" rel="noopener noreferrer"><strong>{plugin.name}</strong></a>
);

const Requirements = ({ requires, requirementsProvided }: RequirementsProps) => {
  const missing = missingRequirements(requires, requirementsProvided);

  return Object.keys(missing).length > 0
    ? (
      <h5>
        Missing requirement(s): {Object.values(missing)
        .map((plugin: Plugin) => <MissingRequirement key={plugin.name} plugin={plugin} />)}
      </h5>
    )
    : null;
};

const View = ({ children, id, title, summary, description, owner, requires, createdAt, requirementsProvided }) => (
  <EntityListItem title={formatTitle(title, id, isMissingRequirements(requires, requirementsProvided))}
                  titleSuffix={summary}
                  description={(
                    <>
                      <Description description={description} owner={owner} createdAt={createdAt} />
                      <Requirements requires={requires} requirementsProvided={requirementsProvided} />
                    </>
                  )}
                  actions={children} />
);

View.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  summary: PropTypes.string,
  description: PropTypes.string,
  owner: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  requires: PropTypes.objectOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
  })).isRequired,
  requirementsProvided: PropTypes.arrayOf(PropTypes.string).isRequired,
};

View.defaultProps = {
  children: null,
  title: 'Unnamed View',
  summary: null,
  description: null,
};

export default withPluginEntities(View, { requirementsProvided: 'views.requires.provided' });
