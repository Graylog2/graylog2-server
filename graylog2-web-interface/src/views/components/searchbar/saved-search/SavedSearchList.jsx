// @flow strict
import React from 'react';
import PropTypes from 'prop-types';
import { browserHistory } from 'react-router';
// $FlowFixMe should be there
import styled, { css, type StyledComponent } from 'styled-components';

import Routes from 'routing/Routes';
import { SavedSearchesStore, SavedSearchesActions } from 'views/stores/SavedSearchesStore';
import type { SavedSearchesState } from 'views/stores/SavedSearchesStore';
import connect from 'stores/connect';
import { Alert, Modal, ListGroup, ListGroupItem, Button } from 'components/graylog';
import { Icon, PaginatedList, SearchForm } from 'components/common';
import View from 'views/logic/views/View';
import { type ThemeInterface } from 'theme';
import ViewLoaderContext from 'views/logic/ViewLoaderContext';

type Props = {
  toggleModal: () => void,
  deleteSavedSearch: (View) => Promise<View>,
  views: SavedSearchesState,
};

type State = {
  selectedSavedSearch?: string,
  query: string,
  page: number,
  perPage: number,
};

const AlertIcon: StyledComponent<{}, ThemeInterface, *> = styled(Icon)(({ theme }) => css`
  margin-right: 6px;
  color: ${theme.colors.variant.primary};
`);

const NoSavedSearches: StyledComponent<{}, ThemeInterface, *> = styled(Alert)`
  clear: right;
  display: flex;
  align-items: center;
`;

const DeleteButton: StyledComponent<{}, ThemeInterface, HTMLSpanElement> = styled.span`
  position: absolute;
  top: 10px;
  right: 10px;
`;

class SavedSearchList extends React.Component<Props, State> {
  static propTypes = {
    toggleModal: PropTypes.func.isRequired,
    deleteSavedSearch: PropTypes.func.isRequired,
    views: PropTypes.object,
  };

  static defaultProps = {
    views: {},
  };

  constructor(props) {
    super(props);

    this.state = {
      selectedSavedSearch: undefined,
      query: '',
      page: 1,
      perPage: 5,
    };
  }

  componentDidMount() {
    this.execSearch();
  }

  execSearch = () => {
    const { query, page, perPage } = this.state;

    SavedSearchesActions.search(query, page, perPage);
  };

  handlePageChange = (page, perPage) => {
    this.setState({ page: page, perPage: perPage }, this.execSearch);
  };

  handleSearch = (query) => {
    this.setState({ query: query, page: 1 }, this.execSearch);
  };

  handleSearchReset = () => {
    this.setState({ query: '', page: 1 }, this.execSearch);
  };

  onLoad = (selectedSavedSearch, loadFunc) => {
    const { toggleModal } = this.props;

    if (!selectedSavedSearch || !loadFunc) {
      return;
    }

    loadFunc(selectedSavedSearch).then(() => {
      browserHistory.push(Routes.pluginRoute('SEARCH_VIEWID')(selectedSavedSearch));
    });

    toggleModal();
  };

  onDelete = (e, selectedSavedSearch) => {
    const { views, deleteSavedSearch } = this.props;
    const { list } = views;

    e.stopPropagation();

    if (list) {
      const viewIndex = list.findIndex((v) => v.id === selectedSavedSearch);

      if (viewIndex < 0) {
        return;
      }

      // eslint-disable-next-line no-alert
      if (window.confirm(`You are about to delete saved search: "${list[viewIndex].title}". Are you sure?`)) {
        deleteSavedSearch(list[viewIndex]).then(() => {
          this.execSearch();
        });
      }
    }
  };

  render() {
    const { views, toggleModal } = this.props;
    const { total, page, perPage = 5 } = views.pagination;
    const { selectedSavedSearch } = this.state;
    const savedSearchList = (views.list || []).map((savedSearch) => {
      return (
        <ViewLoaderContext.Consumer key={savedSearch.id}>
          {(loaderFunc) => (
            <ListGroupItem active={selectedSavedSearch === savedSearch.id}
                           onClick={() => this.onLoad(savedSearch.id, loaderFunc)}>
              {savedSearch.title}
              <span>
                <DeleteButton bsSize="xsmall" bsStyle="danger" onClick={(e) => this.onDelete(e, savedSearch.id)}>
                  <Icon name="trash" ttile="Delete" data-testid={`delete-${savedSearch.id}`} />
                </DeleteButton>
              </span>
            </ListGroupItem>
          )}
        </ViewLoaderContext.Consumer>
      );
    });

    const renderResult = ((views && views.list) && views.list.length > 0)
      ? (
        <PaginatedList onChange={this.handlePageChange}
                       activePage={page}
                       totalItems={total}
                       pageSize={perPage}
                       pageSizes={[5, 10, 15]}>
          <ListGroup>{savedSearchList}</ListGroup>
        </PaginatedList>
      )
      : (
        <NoSavedSearches>
          <AlertIcon name="exclamation-triangle" size="lg" />
          <span>No saved searches found.</span>
        </NoSavedSearches>
      );

    return (
      <Modal show>
        <Modal.Body>
          <SearchForm focusAfterMount
                      onSearch={this.handleSearch}
                      onReset={this.handleSearchReset} />
          {renderResult}
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={toggleModal}>Cancel</Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

export default connect(SavedSearchList, { views: SavedSearchesStore });
