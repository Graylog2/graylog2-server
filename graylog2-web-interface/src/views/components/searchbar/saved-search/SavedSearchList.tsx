/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import type { PaginatedViews } from 'views/stores/ViewManagementStore';
import { SavedSearchesActions } from 'views/stores/SavedSearchesStore';
import { Alert, Modal, ListGroup, ListGroupItem, Button } from 'components/graylog';
import { Icon, PaginatedList, SearchForm, Spinner } from 'components/common';
import View from 'views/logic/views/View';
import ViewLoaderContext from 'views/logic/ViewLoaderContext';
import { ViewManagementActions } from 'views/stores/ViewManagementStore';
import UserNotification from 'util/UserNotification';
import QueryHelper from 'components/common/QueryHelper';

type Props = {
  toggleModal: () => void,
  deleteSavedSearch: (view: View) => Promise<View>,
  activeSavedSearchId: string,
};

const NoSavedSearches = styled(Alert)`
  clear: right;
  display: flex;
  align-items: center;
  margin-top: 15px;
`;

const ListContainer = styled.div`
  margin-top: 15px;
`;

const LoadingSpinner = styled(Spinner)`
  margin-top: 15px;
`;

const DeleteButton = styled.span(({ theme }) => `
  position: absolute;
  margin-right: 10px;
  width: 25px;
  height: 25px;
  right: 0;
  top: 5px;
  text-align: center;
  display: flex;
  align-items: center;
  justify-content: center;
  color: ${theme.colors.gray[60]};
`);

const DEFAULT_PAGINATION = {
  query: '',
  page: 1,
  perPage: 10,
  count: 0,
};

const onLoad = (toggleModal, selectedSavedSearchId, loadFunc) => {
  if (!selectedSavedSearchId || !loadFunc) {
    return;
  }

  loadFunc(selectedSavedSearchId);

  toggleModal();
};

const onDelete = (e, savedSearches, deleteSavedSearch, selectedSavedSearchId) => {
  e.stopPropagation();

  const selectedSavedSearch = savedSearches?.find((savedSearch) => savedSearch.id === selectedSavedSearchId);

  if (savedSearches) {
    // eslint-disable-next-line no-alert
    if (window.confirm(`You are about to delete saved search: "${selectedSavedSearch.title}". Are you sure?`)) {
      deleteSavedSearch(selectedSavedSearch);
    }
  }
};

const _loadSavesSearches = (pagination, setLoading, setPaginatedSavedSearches) => {
  setLoading(true);

  SavedSearchesActions.search(pagination).then((paginatedSavedSearches) => {
    setPaginatedSavedSearches(paginatedSavedSearches);
    setLoading(false);
  }).catch((error) => {
    UserNotification.error(`Fetching saved searches failed with status: ${error}`,
      'Could not retrieve saved searches');
  });
};

const _updateListOnSearchDelete = (perPage, query, setPagination) => ViewManagementActions.delete.completed.listen(() => setPagination({ page: DEFAULT_PAGINATION.page, perPage, query }));

const SavedSearchList = ({ toggleModal, deleteSavedSearch, activeSavedSearchId }: Props) => {
  const [paginatedSavedSearches, setPaginatedSavedSearches] = useState<PaginatedViews | undefined>();
  const [pagination, setPagination] = useState(DEFAULT_PAGINATION);
  const [loading, setLoading] = useState(false);
  const { page, query, perPage } = pagination;
  const { list: savedSearches, pagination: { total = 0 } = {} } = paginatedSavedSearches || {};

  useEffect(() => _loadSavesSearches(pagination, setLoading, setPaginatedSavedSearches), [pagination]);
  useEffect(() => _updateListOnSearchDelete(perPage, query, setPagination), [perPage, query]);

  const handleSearch = (newQuery: string) => setPagination({ ...pagination, query: newQuery, page: DEFAULT_PAGINATION.page });
  const handlePageSizeChange = (newPage: number, newPerPage: number) => setPagination({ ...pagination, page: newPage, perPage: newPerPage });

  return (
    <Modal show>
      <Modal.Body>
        <PaginatedList onChange={handlePageSizeChange}
                       activePage={page}
                       totalItems={total}
                       pageSize={perPage}>
          <SearchForm focusAfterMount
                      onSearch={handleSearch}
                      queryHelpComponent={<QueryHelper entityName="search" commonFields={['id', 'title']} />}
                      topMargin={0}
                      onReset={() => handleSearch('')} />
          {loading && (<LoadingSpinner />)}
          <ListContainer>
            {!loading && total === 0 && (
              <NoSavedSearches>
                No saved searches found.
              </NoSavedSearches>
            )}
            {savedSearches?.length > 0 && (
              <ListGroup>
                {savedSearches.map((savedSearch) => (
                  <ViewLoaderContext.Consumer key={savedSearch.id}>
                    {(loaderFunc) => (
                      <ListGroupItem onClick={() => onLoad(toggleModal, savedSearch.id, loaderFunc)}
                                     header={savedSearch.title}
                                     active={savedSearch.id === activeSavedSearchId}>
                        {savedSearch.summary}
                        <DeleteButton onClick={(e) => onDelete(e, savedSearches, deleteSavedSearch, savedSearch.id)}
                                      role="button"
                                      title={`Delete search ${savedSearch.title}`}
                                      tabIndex={0}>
                          <Icon name="trash-alt" />
                        </DeleteButton>
                      </ListGroupItem>
                    )}
                  </ViewLoaderContext.Consumer>
                ))}
              </ListGroup>
            )}
          </ListContainer>
        </PaginatedList>
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={toggleModal}>Cancel</Button>
      </Modal.Footer>
    </Modal>
  );
};

SavedSearchList.propTypes = {
  toggleModal: PropTypes.func.isRequired,
  deleteSavedSearch: PropTypes.func.isRequired,
};

export default SavedSearchList;
