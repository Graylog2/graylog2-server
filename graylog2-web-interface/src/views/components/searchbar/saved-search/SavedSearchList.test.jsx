// @flow strict
import React from 'react';
import { render, cleanup, fireEvent, wait } from 'wrappedTestingLibrary';
import { browserHistory } from 'react-router';

import Routes from 'routing/Routes';
import View from 'views/logic/views/View';
import ViewLoaderContext from 'views/logic/ViewLoaderContext';

import SavedSearchList from './SavedSearchList';

const createViewsResponse = (count = 1) => {
  const views = [];

  if (count > 0) {
    // eslint-disable-next-line no-plusplus
    for (let i = 0; i < count; i++) {
      views.push(
        View.builder()
          .id(`foo-bar-${i}`)
          .title(`test-${i}`)
          .description('desc')
          .build(),
      );
    }
  }

  return {
    pagination: {
      total: count,
      page: count > 0 ? count : 0,
      perPage: 5,
    },
    list: views,
  };
};

describe('SavedSearchList', () => {
  describe('render the SavedSearchList', () => {
    afterEach(() => {
      cleanup();
    });

    it('should render empty', () => {
      const views = createViewsResponse(0);
      const { baseElement } = render(<SavedSearchList toggleModal={() => {}}
                                                      showModal
                                                      deleteSavedSearch={() => {}}
                                                      views={views} />);

      expect(baseElement).toMatchSnapshot();
    });

    it('should render with views', () => {
      const views = createViewsResponse(1);
      const { baseElement } = render(<SavedSearchList toggleModal={() => {}}
                                                      showModal
                                                      deleteSavedSearch={() => {}}
                                                      views={views} />);

      expect(baseElement).toMatchSnapshot();
    });

    it('should handle toggle modal', () => {
      const onToggleModal = jest.fn();
      const views = createViewsResponse(1);

      const { getByText } = render(<SavedSearchList toggleModal={onToggleModal}
                                                    showModal
                                                    deleteSavedSearch={() => {}}
                                                    views={views} />);

      const cancel = getByText('Cancel');

      fireEvent.click(cancel);

      expect(onToggleModal).toBeCalledTimes(1);
    });

    it('should handle with delete', () => {
      window.confirm = jest.fn(() => true);
      const onDelete = jest.fn(() => {
        return new Promise(() => {});
      });
      const views = createViewsResponse(1);
      const { getByTestId } = render(<SavedSearchList toggleModal={() => {}}
                                                      showModal
                                                      deleteSavedSearch={onDelete}
                                                      views={views} />);
      const deleteBtn = getByTestId('delete-foo-bar-0');

      fireEvent.click(deleteBtn);

      expect(window.confirm).toBeCalledTimes(1);
      expect(onDelete).toBeCalledTimes(1);
    });

    it('should handle with load', () => {
      const onLoad = jest.fn(() => { return new Promise(() => {}); });
      const views = createViewsResponse(1);

      const { getByText } = render(
        <ViewLoaderContext.Provider value={onLoad}>
          <SavedSearchList toggleModal={() => {}}
                           showModal
                           deleteSavedSearch={() => {}}
                           views={views} />
        </ViewLoaderContext.Provider>,
      );
      const listItem = getByText('test-0');

      fireEvent.click(listItem);

      expect(onLoad).toBeCalledTimes(1);
    });
  });

  describe('load new saved search', () => {
    afterEach(() => {
      cleanup();
    });

    it('should change url after load', async () => {
      const onLoad = jest.fn(() => Promise.resolve());

      Routes.pluginRoute = jest.fn((route) => (id) => `${route}:${id}`);
      browserHistory.push = jest.fn();
      const views = createViewsResponse(1);

      const { getByText } = render(
        <ViewLoaderContext.Provider value={onLoad}>
          <SavedSearchList toggleModal={() => {}}
                           showModal
                           deleteSavedSearch={() => {}}
                           views={views} />
        </ViewLoaderContext.Provider>,
      );
      const listItem = getByText('test-0');

      fireEvent.click(listItem);
      await wait(() => {
        expect(browserHistory.push).toBeCalledTimes(1);
        expect(browserHistory.push).toHaveBeenCalledWith('SEARCH_VIEWID:foo-bar-0');
      });
    });
  });
});
