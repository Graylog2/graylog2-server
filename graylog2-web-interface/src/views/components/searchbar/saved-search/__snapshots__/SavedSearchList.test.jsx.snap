// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedSearchList render the SavedSearchList should render empty 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-hidden="true"
    role="dialog"
  >
    <div
      class="fade modal-backdrop in"
    />
    <div
      class="fade Modal-nuree8-0 eNmkew in modal"
      role="dialog"
      style="display: block; padding-right: 0px;"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
          role="document"
        >
          <div
            class="modal-body"
          >
            <div
              class="search"
              style="margin-top: 15px;"
            >
              <form
                class="form-inline"
              >
                <div
                  class="form-group has-feedback"
                >
                  <input
                    autocomplete="off"
                    class="query form-control"
                    id="common-search-form-query-input"
                    placeholder="Enter search query..."
                    spellcheck="false"
                    style="width: auto;"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-group"
                  style="margin-left: 5px;"
                >
                  <button
                    class="Button__StyledButton-c9cbmb-0 hpBclA submit-button btn btn-default"
                    type="submit"
                  >
                    Search
                  </button>
                </div>
                <div
                  class="form-group"
                  style="margin-left: 5px;"
                >
                  <button
                    class="Button__StyledButton-c9cbmb-0 hpBclA reset-button btn btn-default"
                    type="reset"
                  >
                    Reset
                  </button>
                </div>
              </form>
            </div>
            <div
              class="Alert__StyledAlert-sc-1yucn9g-0 gthrMV SavedSearchList__NoSavedSearches-sc-1r3qvm6-1 gmJDqJ alert alert-info"
              role="alert"
            >
              <svg
                class="svg-inline--fa fa-exclamation-triangle"
              />
              <span>
                No saved searches found.
              </span>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="Button__StyledButton-c9cbmb-0 hpBclA btn btn-default"
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`SavedSearchList render the SavedSearchList should render with views 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-hidden="true"
    role="dialog"
  >
    <div
      class="fade modal-backdrop in"
    />
    <div
      class="fade Modal-nuree8-0 eNmkew in modal"
      role="dialog"
      style="display: block; padding-right: 0px;"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
          role="document"
        >
          <div
            class="modal-body"
          >
            <div
              class="search"
              style="margin-top: 15px;"
            >
              <form
                class="form-inline"
              >
                <div
                  class="form-group has-feedback"
                >
                  <input
                    autocomplete="off"
                    class="query form-control"
                    id="common-search-form-query-input"
                    placeholder="Enter search query..."
                    spellcheck="false"
                    style="width: auto;"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-group"
                  style="margin-left: 5px;"
                >
                  <button
                    class="Button__StyledButton-c9cbmb-0 hpBclA submit-button btn btn-default"
                    type="submit"
                  >
                    Search
                  </button>
                </div>
                <div
                  class="form-group"
                  style="margin-left: 5px;"
                >
                  <button
                    class="Button__StyledButton-c9cbmb-0 hpBclA reset-button btn btn-default"
                    type="reset"
                  >
                    Reset
                  </button>
                </div>
              </form>
            </div>
            <div
              class="form-inline page-size"
              style="float: right;"
            >
              <div
                class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 gcTkZc form-group"
              >
                <label
                  class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                  for="page-size"
                >
                  Show:
                </label>
                <span>
                  <select
                    class="FormControl-sc-1amoaox-0 hBmgnF form-control input-sm"
                    id="page-size"
                    label="Show:"
                    name="page-size"
                    placeholder=""
                    type="select"
                  >
                    <option
                      value="5"
                    >
                      5
                    </option>
                    <option
                      value="10"
                    >
                      10
                    </option>
                    <option
                      value="15"
                    >
                      15
                    </option>
                  </select>
                  
                </span>
              </div>
            </div>
            <div
              class="list-group"
            >
              <button
                class="ListGroupItem__StyledListGroupItem-sc-1ky0joo-0 XtBNB list-group-item"
                type="button"
              >
                test-0
                <span>
                  <span
                    class="SavedSearchList__DeleteButton-sc-1r3qvm6-2 hOFiKI"
                  >
                    <svg
                      class="svg-inline--fa fa-trash-alt"
                      data-testid="delete-foo-bar-0"
                    />
                  </span>
                </span>
              </button>
            </div>
            <div
              class="text-center pagination-wrapper"
            />
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="Button__StyledButton-c9cbmb-0 hpBclA btn btn-default"
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
